src/session-pool.js

```diff
@@ -304,57 +192,100 @@ SessionPool.prototype.getSession = function(callback) {
 /**
  * Retrieve the write session.
  *
- * @param {GetWriteSessionCallback} callback Callback function.
+ * @param {GetWriteSessionCallback} [callback] Callback function.
+ * @returns {Promise<Session>}
  */
 SessionPool.prototype.getWriteSession = function(callback) {
-  var pool = this.writePool;
+  var promise = this.borrowSession_(READWRITE);
 
-  if (!pool || !pool.free) {
-    this.getNextAvailableSession_(
-      {
-        write: true,
-      },
-      callback
-    );
-    return;
+  if (!is.fn(callback)) {
+    return promise;
   }
 
-  pool
-    .acquire()
-    .then(function(session) {
-      // because of the way generic-pool handles creation errors, we need to
-      // resolve our create calls with any errors that occur. That being said,
-      // it is possible that this instance of `session` will actually be an
-      // error of some kind.
-      // re: https://github.com/coopernurse/node-pool/issues/175
-      if (session instanceof Error) {
-        pool.destroy(session);
-        callback(session);
-        return;
-      }
+  promise.then(function(session) {
+    callback(null, session, session.txn);
+  }, callback);
+};
 
-      callback(null, session, session.transaction_);
-    })
-    .then(null, callback);
+/**
+ * Checks to see if the pool is full by calculating the current size and the
+ * pending number of sessions being created.
+ *
+ * @return {boolean}
+ */
+SessionPool.prototype.isFull = function() {
+  return this.size() + this.pendingCreates_ === this.options.max;
 };
 
 /**
- * Release a session back into the pool.
+ * Opens the pool, filling it to the configured number of read and write
+ * sessions.
  *
- * @param {Session} session The session to be released.
- * @returns {Promise}
+ * @emits SessionPool#open
+ * @param {function} [callback] The callback function.
+ * @return {Promise}
+ */
+SessionPool.prototype.open = function(callback) {
+  this.isOpen = true;
+  this.listenForEvents_();
+  this.emit('open');
+
+  var neededReadWrite = this.options.writes;
+  var neededReadOnly = this.options.min - neededReadWrite;
+
+  var requests = [];
+  var i;
+
+  for (i = 0; i < neededReadWrite; i++) {
+    requests.push(this.createSessionInBackground_(READWRITE));
```

Comment:
what happens if these creations fail? For eg if the user gave an incorrect database name, all session creations would fail.

Reply:
In the event of a create failure we would emit an error that the Database object would catch and then re-emit said error to the user.

Reply:
I see. But what would happen if they call something like `run_in_transaction` on this database? Will there callback never get invoked or will it get invoked with an error?

Reply:
It should get invoked with an error, basically our flow looks like this

* Is there a session available?
  * Yes! Return it!
* Did the user disable waiting for a session?
  * Yes! Return an error.
* Is the pool full?
  * No! Create a session and return that. If an error occurs while creating, we'll return this error to the user.
  * Yes! Set a timeout (if one is configured) and race between the pool closing, the timeout occurring and a session being released back to the pool.

Reply:
Perfect. Thanks for clarifying.